module/zfs/arc.c

```diff
@@ -4066,25 +4113,108 @@ arc_evict_state(arc_state_t *state, arc_buf_contents_t type, uint64_t spa,
 		multilist_sublist_unlock(mls);
 	}
 
+	evict_arg_t *evarg = kmem_alloc(sizeof (*evarg) * num_sublists,
+	    KM_SLEEP);
 	/*
 	 * While we haven't hit our target number of bytes to evict, or
 	 * we're evicting all available buffers.
 	 */
 	while (total_evicted < bytes) {
 		int sublist_idx = multilist_get_random_index(ml);
+		boolean_t usetskq = zfs_arc_evict_parallel;
 		uint64_t scan_evicted = 0;
 
+		uint64_t left = (bytes == ARC_EVICT_ALL ? bytes :
+		    bytes - total_evicted);
+
+		/*
+		 *  How we scale
+		 *
+		 *  Example 1, # of chunks less than # of tasks.
+		 *  We have:
+		 *  - 4 tasks
+		 *  - 3 chunks
+		 *  - 3 full col
+		 *  - 0 low cols.
+		 *
+		 *  The first low col index is 3.
+		 *  The tasks #0-#2 evict 1 chunk each.
+		 *
+		 *    0 | 1 | 2 | 3 |
+		 *  +===+===+===+===+
+		 *  | x | x | x |   |
+		 *  +---+---+---+---+
+		 *
+		 *  Example 2, # of chunks more than # of tasks.
+		 *  We have:
+		 *  - 4 tasks
+		 *  - 9 chunks
+		 *  - 1 full col
+		 *  - 3 low cols
+		 *
+		 *  The first low col index is 1.
+		 *  The task #0 evicts 3 chunks, the others evict 2 chunks each.
+		 *
+		 *    0 | 1 | 2 | 3 |
+		 *  +===+===+===+===+
+		 *  | x | x | x | x |
+		 *  +---+---+---+---+
+		 *  | x | x | x | x |
+		 *  +---+---+---+---+
+		 *  | x |   |   |   |
+		 *  +---+---+---+---+
+		 */
+
+		/*
+		 * Compute number of tasks to run (n), low col index (k)
+		 * and normal and low bytes per task.
+		 */
+		uint64_t nchunks = ((left - 1) >> MIN_EVICT_PERTASK_SHIFT) + 1;
+		unsigned n = nchunks < num_sublists ? nchunks : num_sublists;
+		uint64_t fullrows = nchunks / n;
+		unsigned lastrowcols = nchunks % n;
+		unsigned k = (lastrowcols ? lastrowcols : n);
+
+		uint64_t bytes_pertask_low =
+		    fullrows << MIN_EVICT_PERTASK_SHIFT;
+		uint64_t bytes_pertask = bytes_pertask_low + (lastrowcols ?
+		    (1 << MIN_EVICT_PERTASK_SHIFT) : 0);
+
 		/*
 		 * Start eviction using a randomly selected sublist,
 		 * this is to try and evenly balance eviction across all
 		 * sublists. Always starting at the same sublist
 		 * (e.g. index 0) would cause evictions to favor certain
 		 * sublists over others.
 		 */
-		for (int i = 0; i < num_sublists; i++) {
+		for (unsigned i = 0; i < n; i++, sublist_idx++) {
 			uint64_t bytes_remaining;
 			uint64_t bytes_evicted;
 
+			/* we've reached the end, wrap to the beginning */
+			if (sublist_idx >= num_sublists)
+				sublist_idx = 0;
+
+			if (usetskq) {
+				uint64_t evict = i < k ? bytes_pertask :
+				    bytes_pertask_low;
+
+				ASSERT3S(n, <=, num_sublists);
```

Comment:
While it is true, I think it is irrelevant here.

Reply:
Right, neither `n` nor `num_sublists` can change in this loop, so I agree there is little sense to keeping this assert around.